sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala

```diff
@@ -0,0 +1,93 @@
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution
+
+import java.util.concurrent.atomic.AtomicLong
+
+import org.apache.spark.SparkContext
+import org.apache.spark.sql.{DataFrame, SQLContext}
+import org.apache.spark.util.Utils
+
+private[sql] object SQLExecution {
+
+  val EXECUTION_ID_KEY = "spark.sql.execution.id"
+
+  private val _nextExecutionId = new AtomicLong(0)
+
+  private def nextExecutionId: Long = _nextExecutionId.getAndIncrement
+
+  /**
+   * Wrap a DataFrame action to track all Spark jobs in the body so that we can connect them with
+   * an execution.
+   */
+  def withNewExecutionId[T](sqlContext: SQLContext, df: DataFrame)(body: => T): T = {
+    val sc = sqlContext.sparkContext
+    val oldExecutionId = sc.getLocalProperty(EXECUTION_ID_KEY)
+    if (oldExecutionId == null) {
+      val executionId = SQLExecution.nextExecutionId
+      sc.setLocalProperty(EXECUTION_ID_KEY, executionId.toString)
+      val r = try {
+        val callSite = Utils.getCallSite()
+        sqlContext.listener.onExecutionStart(
+          executionId, callSite.shortForm, callSite.longForm, df, System.currentTimeMillis())
+        try {
+          body
+        } finally {
+          // Ideally, we need to make sure onExecutionEnd happens after onJobStart and onJobEnd.
+          // However, onJobStart and onJobEnd run in the listener thread. Because we cannot add new
+          // SQL event types to SparkListener since it's a public API, we cannot guarantee that.
+          //
+          // SQLListener should handle the case that onExecutionEnd happens before onJobEnd.
+          //
+          // The worst case is onExecutionEnd may happen before onJobStart when the listener thread
+          // is very busy. If so, we cannot track the jobs for the execution. It seems acceptable.
+          sqlContext.listener.onExecutionEnd(executionId, System.currentTimeMillis())
+        }
+      } finally {
+        sc.setLocalProperty(EXECUTION_ID_KEY, null)
+      }
+      r
+    } else {
+      // Don't support nested `withNewExecution`. This is an example of the nested
+      // `withNewExecution`:
+      //
+      // class DataFrame {
+      //   def foo: T = withNewExecution { something.createNewDataFrame().collect() }
+      // }
+      //
+      // Note: `collect` will call withNewExecution
+      // In this case, only the "executedPlan" for "collect" will be executed. The "executedPlan"
+      // for the outer DataFrame won't be executed. So it's meaningless to create a new Execution
+      // for the outer DataFrame. Even if we track it, since its "executedPlan" doesn't run,
+      // all accumulator metrics will be 0. It will confuse people if we show them in Web UI.
+      //
+      // A real case is the `DataFrame.count` method.
+      throw new IllegalArgumentException(s"$EXECUTION_ID_KEY is already set")
+    }
+  }
+
+  def withExecutionId[T](sc: SparkContext, executionId: String)(body: => T): T = {
```

Comment:
can you add a java doc on when this is called? When do we want to reuse an existing execution ID instead of creating a new one?